compiler/noirc_evaluator/src/ssa/opt/inline_const_brillig_calls.rs

```diff
@@ -0,0 +1,221 @@
+//! This pass tries to inline calls to brillig functions that have all constant arguments.
+use std::collections::{BTreeMap, HashSet};
+
+use acvm::acir::circuit::ErrorSelector;
+use noirc_frontend::{monomorphization::ast::InlineType, Type};
+
+use crate::{
+    errors::RuntimeError,
+    ssa::{
+        ir::{
+            function::{Function, FunctionId, RuntimeType},
+            instruction::{Instruction, InstructionId, TerminatorInstruction},
+            value::{Value, ValueId},
+        },
+        optimize_ssa_after_inline_const_brillig_calls, Ssa, SsaBuilder,
+    },
+};
+
+impl Ssa {
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub(crate) fn inline_const_brillig_calls(mut self, inliner_aggressiveness: i64) -> Self {
+        let error_selector_to_type = &self.error_selector_to_type;
+
+        // Collect all brillig functions so that later we can find them when processing a call instruction
+        let mut brillig_functions = BTreeMap::<FunctionId, Function>::new();
+        for (func_id, func) in &self.functions {
+            if let RuntimeType::Brillig(..) = func.runtime() {
+                let cloned_function = Function::clone_with_id(*func_id, func);
+                brillig_functions.insert(*func_id, cloned_function);
+            };
+        }
+
+        // Keep track of which brillig functions we couldn't completely inline: we'll remove the ones we could.
+        let mut brillig_functions_we_could_not_inline = HashSet::new();
+
+        for func in self.functions.values_mut() {
+            func.inline_const_brillig_calls(
+                &brillig_functions,
```

Comment:
Why not loop through `brillig_functions` to get the IDs of the ones we already shortlisted?

Reply:
Ah because you are trying to inline the brillig function which are called from any function at all.

If brillig function `foo` is called from non-brillig functions `bar` and `baz`, then it is going to be attempted to be optimised twice, with respective constant arguments from both call sites.